packages/contracts-bedrock/scripts/differential-testing/utils.go

```diff
@@ -0,0 +1,155 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/crossdomain"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var UnknownNonceVersion = errors.New("Unknown nonce version")
+
+var L2ToL1MessagePasser = common.HexToAddress("0x4200000000000000000000000000000000000016")
+
+var EmptyTrieHash = common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+
+// checkOk checks if ok is false, and panics if so.
+// Shorthand to ease go's god awful error handling
+func checkOk(ok bool) {
+	if !ok {
+		panic(fmt.Errorf("checkOk failed"))
+	}
+}
+
+// checkErr checks if err is not nil, and throws if so.
+// Shorthand to ease go's god awful error handling
+func checkErr(err error, failReason string) {
+	if err != nil {
+		panic(fmt.Errorf("%s: %s", failReason, err))
+	}
+}
+
+// encodeCrossDomainMessage encodes a versioned cross domain message into a byte array.
+func encodeCrossDomainMessage(nonce *big.Int, sender common.Address, target common.Address, value *big.Int, gasLimit *big.Int, data []byte) ([]byte, error) {
+	_, version := crossdomain.DecodeVersionedNonce(nonce)
+
+	var encoded []byte
+	var err error
+	if version.Cmp(big.NewInt(0)) == 0 {
+		// Encode cross domain message V0
+		encoded, err = crossdomain.EncodeCrossDomainMessageV0(target, sender, data, nonce)
+	} else if version.Cmp(big.NewInt(1)) == 0 {
+		// Encode cross domain message V1
+		encoded, err = crossdomain.EncodeCrossDomainMessageV1(nonce, sender, target, value, gasLimit, data)
+	} else {
+		return nil, UnknownNonceVersion
+	}
+
+	return encoded, err
+}
+
+// hashWithdrawal hashes a withdrawal transaction.
+func hashWithdrawal(nonce *big.Int, sender common.Address, target common.Address, value *big.Int, gasLimit *big.Int, data []byte) (common.Hash, error) {
+	// Pack withdrawal
```

Comment:
This could be useful here: https://github.com/ethereum-optimism/optimism/blob/35d4a37c2b9a59e9bf16fa4c5e22283f2348c3d2/op-chain-ops/crossdomain/withdrawal.go#L122